src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -160,28 +161,39 @@ private void CheckInvariants()
         var projectId = newProjectState.Id;
 
         var newDependencyGraph = newSolutionState.GetProjectDependencyGraph();
-        var newTrackerMap = CreateCompilationTrackerMap(projectId, newDependencyGraph);
+        var newTrackerMap = CreateCompilationTrackerMap(projectId, newDependencyGraph, modifyTrackerMapInfo, (translate, forkTracker, projectId, newProjectState));
 
-        // If we have a tracker for this project, then fork it as well (along with the
-        // translation action and store it in the tracker map.
-        if (newTrackerMap.TryGetValue(projectId, out var tracker))
-        {
-            newTrackerMap = newTrackerMap.Remove(projectId);
+        return this.Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap);
 
-            if (forkTracker)
+        static bool modifyTrackerMapInfo(
+            Dictionary<ProjectId, ICompilationTracker> trackerMap,
+            (TranslationAction? translate, bool forkTracker, ProjectId projectId, ProjectState newProjectState) arg)
+        {
+            // If we have a tracker for this project, then fork it as well (along with the
+            // translation action and store it in the tracker map.
+            if (trackerMap.TryGetValue(arg.projectId, out var tracker))
             {
-                newTrackerMap = newTrackerMap.Add(projectId, tracker.Fork(newProjectState, translate));
+                trackerMap.Remove(arg.projectId);
+
+                if (arg.forkTracker)
+                    trackerMap.Add(arg.projectId, tracker.Fork(arg.newProjectState, arg.translate));
+
+                return true;
             }
-        }
 
-        return this.Branch(
-            newSolutionState,
-            projectIdToTrackerMap: newTrackerMap);
+            return false;
+        }
     }
 
-    private ImmutableDictionary<ProjectId, ICompilationTracker> CreateCompilationTrackerMap(ProjectId changedProjectId, ProjectDependencyGraph dependencyGraph)
+    private ImmutableSegmentedDictionary<ProjectId, ICompilationTracker> CreateCompilationTrackerMap<TArg>(
+        ProjectId changedProjectId,
+        ProjectDependencyGraph dependencyGraph,
+        Func<Dictionary<ProjectId, ICompilationTracker>, TArg, bool> modifyNewTrackerInfo,
```

Comment:
nit consider doc'ing in some form.  perhaps saying that the `bool` indicates if it actually modified or not.